### ISSUE TITLE
Entering Play Mode randomly takes longer to load sometimes

### ISSUE THREAD

**Issue by KXI-System** [YarnSpinnerTool/YarnSpinner-Unity#188]
_(This is a post documenting findings of this issue on Discord)_

There is this elusive bug where some users are finding increased loading times when entering Play Mode _sometimes_. The increased load times appear inconsistently, and seemingly appears after some time before then resolving itself.
Users who have added in the Sample files have found that removing them decreases the load times for a while, but then the load times increase again after a while.

On Discord, 2 users confirmed to have the same issue. Both have the "**Enter Play Mode**" option enabled as well.

---

Discord Logs:

@/Magna on Discord

> Unity version: Unity (2021.3.0f1) 
> I have this problem where sometimes it takes a lot to load, from 1 min to 2, I'm not using Yarn spinner yet in the project, only have it installed
> ![unknown](https://user-images.githubusercontent.com/13906174/179379612-0db70040-b141-4ad8-b2c1-00d8535fdb12.png)
> Packages in project
> ![unknown](https://user-images.githubusercontent.com/13906174/179379620-d85ef831-2a9d-4597-b33e-b2dab935583c.png)

> Btw Idk about versus but I have this enabled
> ![unknown](https://user-images.githubusercontent.com/13906174/179379658-5046cc21-b503-44bd-be8f-15ce9771fcd3.png)

[link to message](https://discord.com/channels/754171172693868585/768556555644829696/997083487070859334)

@/versus_ on Discord

> (Sorry to be meddling in your conversation, but I do have the samples imported since my very first tests to get to know Yarn and it is still from time to time taking over 3 minutes)

> [in response to removing example files]
> I am still getting slightly slow times, but in any case nowhere near the 3 and a half minutes I have been getting

> Sorry to be a bummer, but it happenned again, even after removing the Samples
> This time I did absolutely nothing, just Alt+Tabbed to the web browser to check the FMod API, went back to Unity and hit Play

[link to message](https://discord.com/channels/754171172693868585/768556555644829696/997085433857069096)

**Comment by Garazbolg** [comment]
We are having a similar issue in our project. At first we tried it in a small demo project and everything worked fine. But when we migrated to our real project (which is not the biggest but still big) we started seeing long loading to get into play mode.

After some investigating I figured it was coming from the YarnPreventPlayMode class that would look for any yarn errors and "prevent play mode" if there was any. But in order to get that information it uses GetAllAssetsOf() which uses AssetDatabase.FindAssets.

AssetDatabase.FindAssets works on the whole project, but because it's all indexed it works pretty fast. When we're looking for only the YarnProject it's fine because it only returns YarnProjects which you shouldn't have too many of. But the type for .yarn files is TextAsset which is the same type as EVERY C# script in your project. 

And suddenly you're sorting through thousands of files multiple times right before starting playmode. There has to be a better way.

For our project I've limited the search to a specific folder without any C# scripts in it, which decreases load time considerably.
That's just a quick patch, do you have any plans for a better/long-term solution ?

**Comment by Cdddo** [comment]
I have encountered this issue constantly.

One of the things I did is disable Search All Assemblies in my YarnProject, as it becomes slow when it is compiling a large project. However the problem persists with YarnPreventPlayMode.

Reading @Garazbolg comment above makes me think... if the YarnProject already has a reference to all the .yarn files, what is the need to *search* for them? Can't they just be grabbed from the YarnProjects?

**Comment by Bryan-Legend** [comment]
Try closing all un-needed unity windows. Some of them can take a while, like the Package Manager tab.

**Comment by Appleguysnake** [comment]
Also have this problem, it's not a major issue although obviously annoying. Happy to try and look out for or test potential causes to narrow it down.

**Comment by Greewi** [comment]
We have the issue on a very large project and it delay the play mode launch more than a minute for some of our developpers. Is there any workaround ?

**Comment by Bryan-Legend** [comment]
Try using this tool to get a call stack on what is taking so long:
https://forum.unity.com/threads/introducing-the-editor-iteration-profiler.908390/

**Comment by Cdddo** [comment]
I fixed this problem by changing YarnEditorUtility.cs to filter the query in line 226 to search only in specific folders instead of the whole unity assets database, which in my project is pretty large (therefore slow), e.g.:

```
=> AssetDatabase.FindAssets(filterQuery, new[] { "Assets/_PROJECT/Dialogue" })
```

It would be better to have the ability to configure an array of folders in the YarnProject inspector...

**Comment by rob5300** [comment]
We also experience this issue. I think enforcing no file errors when playing in this way is a bad idea, seeing as for large projects it takes way too long to process.

Ideally this is optimized, allow the functionality to be disabled, or these errors are processed at another time (import, runtime load).

**Comment by leohilbert** [comment]
Was just profiling my code because entering Playmode takes longer and longer and found out that 2sec of it is designated to this YarnSpinner check. This quickly adds up over time.
I will manually copy in the YarnSpinner project, so I can make the adjustment that @Cdddo mentioned, but I'd be great to have a configuration of some kind.

**Comment by desplesda** [comment]
We believe that this is now fixed as of the upcoming v2.3 release. I'll close this issue here; because the underlying code has changed quite a bit, we'll consider any reoccurrence of behaviour like this to be a new issue.

Thanks to everyone for contributing diagnostics that helped track down this problem!
